src/crypto_provider.rs

```diff
@@ -0,0 +1,270 @@
+use libc::size_t;
+use std::slice;
+use std::sync::Arc;
+
+use rustls::crypto::ring;
+use rustls::crypto::CryptoProvider;
+use rustls::SupportedCipherSuite;
+
+use crate::cipher::rustls_supported_ciphersuite;
+use crate::{
+    arc_castable, box_castable, ffi_panic_boundary, free_arc, free_box, rustls_result,
+    set_arc_mut_ptr, set_boxed_mut_ptr, to_arc_const_ptr, to_boxed_mut_ptr, try_clone_arc,
+    try_mut_from_ptr, try_mut_from_ptr_ptr, try_ref_from_ptr, try_ref_from_ptr_ptr, try_slice,
+    try_take,
+};
+
+box_castable! {
+    /// A `rustls_crypto_provider` builder.
+    pub struct rustls_crypto_provider_builder(Option<CryptoProviderBuilder>);
+}
+
+/// A builder for customizing a `CryptoProvider`. Can be used to install a process-wide default.
+#[derive(Debug)]
+pub struct CryptoProviderBuilder {
+    base: Arc<CryptoProvider>,
+    cipher_suites: Vec<SupportedCipherSuite>,
+}
+
+impl CryptoProviderBuilder {
+    fn build_provider(self) -> CryptoProvider {
+        let cipher_suites = match self.cipher_suites.is_empty() {
+            true => self.base.cipher_suites.clone(),
+            false => self.cipher_suites,
+        };
+
+        // Unfortunately we can't use the `..` syntax to fill in the rest of the provider
+        // fields, because we're working with `Arc<CryptoProvider>` as the base,
+        // not `CryptoProvider`.
+        CryptoProvider {
```

Comment:
Perhaps `..self.base.unwrap_or_clone()`?

Reply:
A good idea, but unfortunately [`Arc::unwrap_or_clone()`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.unwrap_or_clone) was only stabilized in 1.76+ and this crate's MSRV is 1.64.

I'll add a TODO for a future cleanup opportunity.